Distinguish projected secret, configmap and downward api tests

msau42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
* Split up projected.go into multiple files: projected_secret/configmap/downwardapi/combined.go
* Added separate Context() with secret, configmap and downward api for projected tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68904

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

msau42 commented:
Ok, the diff looks ridiculous, but all I did was add 3 Context() blocks!

msau42 commented:
/assign @BenTheElder

msau42 commented on test/conformance/testdata/conformance.txt:
```diff
@@ -112,15 +112,15 @@ test/e2e/common/projected.go: "should be consumable from pods in volume as non-r
 test/e2e/common/projected.go: "should be consumable from pods in volume with mappings as non-root"
 test/e2e/common/projected.go: "updates should be reflected in volume"
-test/e2e/common/projected.go: "optional updates should be reflected in volume"
+test/e2e/common/projected.go: "configMap with updates should be reflected in volume"
```
oops

BenTheElder commented:
Thank you! cc @spiffxp

BenTheElder commented on test/conformance/testdata/conformance.txt:
```diff
@@ -112,15 +112,15 @@ test/e2e/common/projected.go: "should be consumable from pods in volume as non-r
 test/e2e/common/projected.go: "should be consumable from pods in volume"
 test/e2e/common/projected.go: "should be consumable from pods in volume with defaultMode set"
 test/e2e/common/projected.go: "should be consumable from pods in volume as non-root"
 test/e2e/common/projected.go: "should be consumable from pods in volume with mappings"
```
this one is still duplicated AFAICT (see line 112 currently)

msau42 replied:
I made the full test case name unique, by adding in another Context().  However, the string inside the It() is still duplicated.

Do we need the It() strings to be unique?

msau42 replied:
the other option I could do is to split out the tests into 3 different files: projected_secret, projected_configmap, projected_downwardapi

BenTheElder replied:
The full identifier needs to be unique, as long as that's the case. I think you're right and that this will already work but it might be worth splitting up anyhow?

BenTheElder commented on test/conformance/testdata/conformance.txt:
```diff
@@ -112,15 +112,15 @@ test/e2e/common/projected.go: "should be consumable from pods in volume as non-r
-test/e2e/common/projected.go: "optional updates should be reflected in volume"
+test/e2e/common/projected.go: "updates should be reflected in volume"
 test/e2e/common/projected.go: "should be consumable from pods in volume"
 test/e2e/common/projected.go: "should be consumable from pods in volume with defaultMode set"
```
this one as well (see line 110 currently)

msau42 replied:
Ok after splitting them into separate files, I am still seeing some duplicates.  So far it looks like at least one of the duplicate tests is a superset of the other.  Can I remove conformance tests?

BenTheElder replied:
I'm not sure what the rules for removing conformance tests are, They're expected to not change post-release but I'm not sure about during a release.

@spiffxp might be able to comment on this.

BenTheElder replied:
er and yeah splitting the files wouldn't be enough I think, they'd need split-file-matching contexts I think? the generated `[sig-storage] Projected should be consumable from pods in volume with mappings [NodeConformance] [Conformance]` entry needs to be unique

msau42 replied:
I changed the top level Describe() of each new file to be unique

spiffxp replied:
maybe we need to update the walker that generates this list, it strips out a lot

msau42 commented on test/e2e/common/projected_configmap.go:
```diff
@@ -0,0 +1,663 @@
+	   Testname: Projected Volume, ConfigMap, update
+	   Description: A Pod is created with projected volume source ‘ConfigMap’ to store a configMap and performs a create and update to new value. Pod MUST be able to create the configMap with value-1. Pod MUST be able to update the value in the confgiMap to value-2.
+	*/
+	framework.ConformanceIt("updates should be reflected in volume [NodeConformance]", func() {
```
This test is a subset of the next test.  I think we can remove this test.

smarterclayton replied:
So optional and non optional projected volumes have exactly the same behavior?

msau42 replied:
Ah I didn't realize there was actually an `Optional` field in the API.  I will add the test back.  Maybe we should add one more test case to test pod failure when optional == false

msau42 replied:
One more observation: non-optional is not tested for secrets either.  So some future areas to improve coverage:
- non-optional for secrets
- non-optional should fail for secrets when key doesn't exist
- non-optional should fail for configmaps when key doesn't exist

spiffxp commented:
/area conformance

spiffxp commented:
/kind cleanup
/priority important-soon

spiffxp commented:
/retest

msau42 commented:
Updated and squashed:
* Separated out tests into projected_configmap, projected_secret, projected_downwardapi
* Removed one duplicate projected_configmap test

BenTheElder commented:
Thanks so much for looking at this! Fixing this will make it viable to get ~correct results faster by running the conformance suite first in parallel for non-`[Serial]` and then in serial for the handful of serial tests.

msau42 commented:
/test pull-kubernetes-integration

spiffxp commented:
/lgtm
I don't think this needs a milestone anymore, I suspect the tide context is out of date
/assign @bgrant0607 @smarterclayton 
for the /approve

bgrant0607 commented:
With changes and a file move in the same commit, it's no longer possible to see what the changes are through the github UI

bgrant0607 commented on test/conformance/testdata/conformance.txt:
```diff
@@ -106,34 +106,33 @@ test/e2e/common/pods.go: "should be submitted and removed"
-test/e2e/common/projected.go: "should provide container's memory request"
-test/e2e/common/projected.go: "should provide node allocatable (cpu) as default cpu limit if the limit is not set"
-test/e2e/common/projected.go: "should provide node allocatable (memory) as default memory limit if the limit is not set"
 test/e2e/common/projected.go: "should project all components that make up the projection API"
```
Why wasn't this one moved?

msau42 replied:
This tests projected secrets, configmaps and downward API combined in one pod.

bgrant0607 commented:
Was any attempt made to preserve history? What does git blame show?

https://stackoverflow.com/questions/3887736/keep-git-history-when-splitting-a-file

msau42 commented:
I'll split this into two commits

msau42 committed:
Keep related test cases and delete the rest

msau42 committed:
Update test suite names

msau42 committed:
Update build and conformance test list

msau42 committed:
Break up projected.go into projected_combined/configmap/secret/downwardapi

msau42 commented:
Split up into multiple commits:
* Copy files and remove original
* Remove content from new files
* Modify test suite names
* Remove duplicate test
* Update conformance test list

Confirmed that git blame retains history on all the new files.

msau42 commented on test/conformance/testdata/conformance.txt:
```diff
@@ -144,6 +143,33 @@ test/e2e/common/secrets_volume.go: "should be consumable from pods in volume wit
 test/e2e/common/secrets_volume.go: "should be able to mount in a volume regardless of a different secret existing with same name in different namespace"
 test/e2e/common/secrets_volume.go: "should be consumable in multiple volumes in a pod"
 test/e2e/common/secrets_volume.go: "optional updates should be reflected in volume"
+test/e2e/common/tmp/projected.go: "should project all components that make up the projection API"
```
oops remove

msau42 commented:
/test pull-kubernetes-integration

msau42 commented:
/test pull-kubernetes-integration

msau42 commented:
this is ready for review. PTAL @spiffxp

msau42 commented:
/test pull-kubernetes-integration

msau42 commented:
Put the non-optional update configmap test case back.  

So this PR is strictly:
* split up projected.go file into each volume type
* rename test suite with the specific volume type to avoid duplicate test case names

spiffxp commented:
/retest
... weird, why are there two different commit shas in those failure links

msau42 commented:
/retest

msau42 commented:
/retest

spiffxp commented:
/retest

smarterclayton commented:
/approve

msau42 commented:
ping @spiffxp for review again

spiffxp commented:
/lgtm
